### PR TITLE
Stub usage report implementation

### PR DESCRIPTION
**Pull request by jaydeepk**
**What**:

Implemented the generation of a 'Stub Usage Report' json file.
The report will be automatically generated once the stub process is stopped.
The report will list all the endpoints for which the stub returned a successful response (in a format similar to the API Coverage Report).

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [x] Tests
- [ ] Sonar Quality Gate

### CODE REVIEW

File: core/src/main/kotlin/in/specmatic/core/HttpResponse.kt [hunk 1]
```diff
@@ -35,6 +35,8 @@ data class HttpResponse(
     fun specmaticResultHeaderValue(): String =
         this.headers.getOrDefault(SPECMATIC_RESULT_HEADER, "success")
 
+    fun isSuccessful(): Boolean = specmaticResultHeaderValue() == "success"
```
Comment (harikrishnan83) [inline core/src/main/kotlin/in/specmatic/core/HttpResponse.kt]: @joelrosario is there another way to check if a response is a success response? also just curious if the result header value is a dependable way of checking if a response is a success response. cc @jaydeepk

File: core/src/main/kotlin/in/specmatic/stub/HttpStub.kt [hunk 2]
```diff
@@ -427,6 +459,40 @@ class HttpStub(
 
     init {
         server.start()
+        extractALlEndpoints()
```
Comment (joelrosario) [inline core/src/main/kotlin/in/specmatic/stub/HttpStub.kt]: I think this method can be called when _allEndpoints is initialised, and then _allEndpoints can become immutable

File: core/src/main/kotlin/in/specmatic/core/HttpResponse.kt [hunk 3]
```diff
@@ -35,6 +35,8 @@ data class HttpResponse(
     fun specmaticResultHeaderValue(): String =
         this.headers.getOrDefault(SPECMATIC_RESULT_HEADER, "success")
```
Comment (joelrosario) [inline core/src/main/kotlin/in/specmatic/core/HttpResponse.kt]: Should we not default to "failure"? If the header does not exist it means failure.

File: core/src/main/kotlin/in/specmatic/core/HttpResponse.kt (hunk 1 above)
Comment (joelrosario) [reply core/src/main/kotlin/in/specmatic/core/HttpResponse.kt]: I think there's a way we can use the type system to signal success or failure.

The code is structured to return a stubbed response. However it's possible to wrap the HttpStubResponse object in another object that contains the behaviour for logging (or not logging as the case may be).

The most expeditious way would be to create a wrapper function around HttpStub/getHttpResponse (or alter it's signature), then introduce the wrapper object in getHttpResponse at line 624, and in fakeHttpResponse at lines 756 and 758.

cc @jaydeepk

**Review by joelrosario — CHANGES_REQUESTED** [review]
I've left some comments.

File: core/src/main/kotlin/in/specmatic/stub/HttpStub.kt [hunk 4]
```diff
@@ -65,9 +67,21 @@ class HttpStub(
...
 
+    companion object {
+        const val JSON_REPORT_PATH = "./build/reports/specmatic"
+        const val JSON_REPORT_FILE_NAME = "stub_usage_report.json"
+    }
+
     private val threadSafeHttpStubs = ThreadSafeListOfStubs(_httpStubs.filter { it.stubToken == null }.toMutableList())
     private val threadSafeHttpStubQueue = ThreadSafeListOfStubs(_httpStubs.filter { it.stubToken != null }.reversed().toMutableList())
 
+    private val _logs: MutableList<StubEndpoint> = mutableListOf()
```
Comment (joelrosario) [inline core/src/main/kotlin/in/specmatic/stub/HttpStub.kt]: This data structure needs to be made thread-safe

**Review by joelrosario — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/in/specmatic/core/HttpResponse.kt (hunk 3 above)
Comment (joelrosario) [reply core/src/main/kotlin/in/specmatic/core/HttpResponse.kt]: This is also used elsewhere in the code in the context of which this behaviour is correct. Closing this.

File: core/src/main/kotlin/in/specmatic/stub/HttpStub.kt (hunk 4 above)
Comment (jaydeepk) [reply core/src/main/kotlin/in/specmatic/stub/HttpStub.kt]: Done

File: core/src/main/kotlin/in/specmatic/stub/HttpStub.kt [hunk 5]
```diff
@@ -65,9 +67,21 @@ class HttpStub(
...
 
+    companion object {
+        const val JSON_REPORT_PATH = "./build/reports/specmatic"
+        const val JSON_REPORT_FILE_NAME = "stub_usage_report.json"
+    }
+
     private val threadSafeHttpStubs = ThreadSafeListOfStubs(_httpStubs.filter { it.stubToken == null }.toMutableList())
     private val threadSafeHttpStubQueue = ThreadSafeListOfStubs(_httpStubs.filter { it.stubToken != null }.reversed().toMutableList())
 
+    private val _logs: MutableList<StubEndpoint> = Collections.synchronizedList(ArrayList())
```
Comment (joelrosario) [inline core/src/main/kotlin/in/specmatic/stub/HttpStub.kt:78]: There should be a test for this. Have say 10 threads running simultaneously making requests, and we should have 10 logs in this list.

**Review by joelrosario — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/in/specmatic/core/HttpResponse.kt (hunk 1 above)
Comment (harikrishnan83) [reply core/src/main/kotlin/in/specmatic/core/HttpResponse.kt]: @jaydeepk @joelrosario okay, if this is dependable way of checking if the response is a success then looks good.

File: core/src/main/kotlin/in/specmatic/stub/HttpStub.kt (hunk 5 above)
Comment (jaydeepk) [reply core/src/main/kotlin/in/specmatic/stub/HttpStub.kt:78]: Sure, I have added the test.

**Review by joelrosario — APPROVED** [review]
Looks good to me, thanks @jaydeepk

**Pull request merged**